### ISSUE TITLE
Support different set of assets depending on --flavor xxx

### ISSUE THREAD

**Issue by zoechi** [flutter/flutter#21682]
I want to create different flavors of the app and they differ (among other things) on the used assets (icons, backgrounds, fonts, logos, ...)

Currently this seems to require to modify `pubspec.yaml` before running `flutter build`
but that is cumbersome.

I'd prefer something like
```
flutter_assets.yaml # shared assets
flutter_assets_chocolate.yaml # additional assets for chocolate flavor
flutter_assets_strawberry.yaml # additional assets for strawberry flavor
```
where the shared assets and the assets where the file suffix matches the `--flavor xxx` value are merged and the result used for `flutter build`.

**Comment by sestegra** [comment]
@zoechi 
Shared assets could be defined in `pubspec.yaml`. What do you think ?

**Comment by zoechi** [comment]
@sestegra I don't think it's a good idea to define them in `pubspec.yaml`, but as far as I am aware there are people drawing plans about that topic with better insight already than I have.

**Comment by sestegra** [comment]
See my proposal #25487

**Comment by sestegra** [comment]
@zoechi Sample project is available here
https://github.com/sestegra/flutter_assets_flavors

**Comment by yousifk** [comment]
@sestegra 
I check your project,that all asset  adds to apk becuase 

> Bundling of package assets
> If the desired asset is specified in the pubspec.yaml file of the package, it is bundled automatically with the application. In particular, assets used by the package itself must be specified in its pubspec.yaml.
> 
> A package can also choose to have assets in its lib/ folder that are not specified in its pubspec.yaml file. In this case, for those images to be bundled, the application has to specify which ones to include in its pubspec.yaml. For instance, a package named fancy_backgrounds could have the following files:
> 
> content_copy
>   .../lib/backgrounds/background1.png
>   .../lib/backgrounds/background2.png
>   .../lib/backgrounds/background3.png
> To include, say, the first image, the pubspec.yaml of the application should specify it in the assets section:
> 
> content_copy
> flutter:
>   assets:
>     - packages/fancy_backgrounds/backgrounds/background1.png
> The lib/ is implied, so it should not be included in the asset path.

[https://flutter.dev/docs/development/ui/assets-and-images#from-packages](https://flutter.dev/docs/development/ui/assets-and-images#from-packages)

SceenShot about analysis apk
[https://imgur.com/2HmGHC1](https://imgur.com/2HmGHC1)

**Comment by sestegra** [comment]
@yousifk  Good remark, but I won't have time to update my pull request #25487 to support further versions. In fact, this pull request only supports Flutter 1.0.0.

Moreover, I guess that Flutter team wouldn't integrate it in short terms.
I suggest you to develop some scripts to update `pubspec.yaml` when compiling different flavors and/or targets.

**Comment by yousifk** [comment]
@sestegra 
Thanks so much ... I do it .bat script 
I will share it at github

**Comment by v7lin** [comment]
any update?

**Comment by Liujingguang** [comment]
#41094 
This PR is mainly for the specific loading of resources for the flavor, and does not consider the factors of platform and mode. If you consider platform and mode, the idea is similar to the PR. The resources that need to be distinguished according to flavor/platform/mode are uniformly placed in a folder, and the name of the folder is specified by --flavor-assets-dir (or other parameters), so that it allows you to filter specific resources based on the flavor/platform/mode information in this folder.

**Comment by jonahwilliams** [comment]
@Hixie do you recall the format you described several weeks (months) ago?

**Comment by PandaGeek1024** [comment]
any updates on this?

**Comment by Hixie** [comment]
@jonahwilliams not sure to what you are referring, can you remind me of the context?

**Comment by jonahwilliams** [comment]
It was an update to the asset scheme to support platform/mode/flavor specific assets. something like:

```
flutter:
  assets:
    - foo.png
    - bar.png
    ios:
      - fizz.png
```
  
We talked about it briefly in discord a while back, but I wasn't able to find the specific conversation

**Comment by amreniouinnovent** [comment]
@zoechi  I wonder if there is an issue for tracking different dependencies based on the flavors.
Use case: Flavor A using a fingerprint SDK, Flavor B using NFC but not fingerprint

**Comment by lucasdidur** [comment]
What is the next step for this feature?

**Comment by jonahwilliams** [comment]
@lucasdidur no one is currently working on this feature (see assignee and milestone). We would accept + review contributions that added this capability. I would note that this feature and testing likely to be fairly involved.

**Comment by v7lin** [comment]
i have more than 10 flavors!!!

**Comment by amreniouinnovent** [comment]
@v7lin, For now, use a script to rename your asset folder

**Comment by v7lin** [comment]
> @v7lin, For now, use a script to rename your asset folder while building

flavors have different dependency and dart code

**Comment by giaur500** [comment]
Any news? What if I have 20 flavors and 50 MB of assets for each flavor. Bundle all assets from all flavors into single apk is not acceptable.

**Comment by NaikSoftware** [comment]
@giaur500 as work around you can use build script something like this
```#!/bin/sh
cp pubspec.yaml pubspec.yaml.bak
cat pubspec-web.yaml >> pubspec.yaml
flutter build web --dart-define=FLUTTER_WEB_USE_SKIA=true --release
rm pubspec.yaml
mv pubspec.yaml.bak pubspec.yaml```

**Comment by giaur500** [comment]
So you mean separate publspec.yaml for each flavor. Well, still not as good, as I need to handle all dependencies in all yaml files and keep all of them up to date, when bump plugin version, I need to edit all files

**Comment by ckdgit** [comment]
#41094
I think this is a solution similar to android

**Comment by giaur500** [comment]
Yes, but as far as I understand, it's rejected

**Comment by freewheelnat** [comment]
@giaur500 You don't need to maintain several pubspec files in your codebase, you just need an extra step in your CI script. 

- checkout your repo
- remove assets file not necessary for the flavor you are building ( you can physically remove them from assets folder, or remove them from being listed in your pubspec file)
- build your flavor

The second step means you need structure for your assets folder to reflect your flavors, eg assets/common (for common files), assets/flavorA, assets/flavorB etc but I assume you already have those to load them properly in your app.

If you don't have CI, you can integrate with Circle CI for free.

**Comment by ckdgit** [comment]
It works.
but if you don't have CI,you can customize the flutter tools,cut assets from other flavor. in one yaml file that you can handle all dependencies for all flavor assets.
in short,all depends ,then Cutting during packaging

**Comment by giaur500** [comment]
As I understand, I need separate yaml file to build each flavor and build script will rename before build, or build script needs to edit yaml before build. That's not very handy in my opinion. Is there any way to store all resoures definitions outside publspec.yaml, in separate file? As far as I know, there's no any way to include anything to yaml. I want to separate resoures definitions and dependecies. Dependecies may change very often and they will be always the same for all flavors, unlike resoures

**Comment by EinatK** [comment]
As far as I can tell, there are many closed issues referencing this problem and they all say it's a good idea and a must have for flutter but there is no real progress on adding official support ?
Is there a timeline on when this would be added ? I have over 50 flavors, loading all the assets or maintaining 50+ pubspec files is not an option. I would love to hear if any progress is being made here. 
Thanks

**Comment by giaur500** [comment]
Indeed, I agree.

**Comment by rockingdice** [comment]
Need solutions to this.

**Comment by jonahwilliams** [comment]
The Flutter framework and tooling is completely open source. If you really need this feature, then consider writing a PR to add it and we'll make sure it gets reviewed. I realize that is not really the answer anyone wants to hear, but the fact of the matter is that we have far more feature requests than we have time to implement, and this one has not been prioritized (see the lack of milestone and low priority issue labels).

**Comment by giaur500** [comment]
Well, maybe it should get higher priority then, because many users keep asking, so it seems to be important. Not all people asking here are able to create PR.

**Comment by rockingdice** [comment]
Before someone is willing to create a PR, I think it should be discussed how to do it is the best way from an official's insight. We don't have to waste our time too.

**Comment by giaur500** [comment]
I agree, have we got any idea that could be considered as correct? Or any suggestions what way is going to be accepted

**Comment by renanyoy** [comment]
maybe adding sections "flavor1:"  "flavor2:"  in the global scope.. with same content than the "flutter:" section. and in the flutter command line just add "-flavor flavor1" or flavor2 or both

**Comment by giaur500** [comment]
From my point of view, any solution that does not force to edit publspec.yaml beforu build is ok. But we need some way that can be accepted.

**Comment by renanyoy** [comment]
maybe restrict flavor names at things like flutter.ios:    flutter.chocolate:  etc..

**Comment by EinatK** [comment]
Why not allow merging with pubspec.yaml per flavor ? i.e. pubspec_flavor.yaml will merge into the main pubspec.yaml, thus having one file that's shared across all flavors and only add / override if there is a flavor specific one (similar to the android manifests)

**Comment by albert0m** [comment]
3 years and still no progress on this basic feature... using scripts is not the ideal solution. Too messy.

**Comment by giaur500** [comment]
Indeed. As a workaround in my case: scripts to create/overwrite symlink to required assets directory. In android studio, scripts are possible to be executed automatically when switching configurations.

But yes, that sucks. Still.

**Comment by talregev** [comment]
@jonahwilliams
> The Flutter framework and tooling is completely open source. If you really need this feature, then consider writing a PR to add it and we'll make sure it gets reviewed. I realize that is not really the answer anyone wants to hear, but the fact of the matter is that we have far more feature requests than we have time to implement, and this one has not been prioritized (see the lack of milestone and low priority issue labels).

There is already publish a solution: 
https://github.com/flutter/flutter/pull/25487
https://github.com/sestegra/flutter_assets_flavors
If this solution is not good enough, please take it and improve it.

**Comment by jstorch** [comment]
> @jonahwilliams
> 
> > The Flutter framework and tooling is completely open source. If you really need this feature, then consider writing a PR to add it and we'll make sure it gets reviewed. I realize that is not really the answer anyone wants to hear, but the fact of the matter is that we have far more feature requests than we have time to implement, and this one has not been prioritized (see the lack of milestone and low priority issue labels).
> 
> There is already publish a solution: #25487 https://github.com/sestegra/flutter_assets_flavors If this solution is not good enough, please take it and improve it.

@jonahwilliams
- How was this a published solution?  Did you look at it?  https://github.com/sestegra/flutter_assets_flavors required patching a 2 year old version of the flutter framework. (1-1/2 when you commented) the problem is, Flavors are 1/2 baked in with no documentation or roadmap.  The gradle build process supports it but asset management within the yaml process is severely lacking.  I would **love** to take this on and get it on my resume (omg, I would love this) but I have already lost 2 weeks trying to get this working only to finally realize I just can’t and the only way is either ubur apk or create a series of complicated scripts and and spend another 2 weeks getting the rest the team and the build server working again.  Search the web, now that Flutter has massive traction and releasing to multiple environments works great, this is an issue.   How can we get this bumped up? 

Thank you,
Jon

**Comment by KPatel91** [comment]
Agreed with @jstorch above - Flavors are integral to our product as we have a whitelabelled application and have transitioned to Flutter, but better support for them would be welcome. As we now have 10 different apps that's a lot of unnecessary assets!

**Comment by KPatel91** [comment]
We desperately need something like this - we use flutter to build a core app that has slightly bespoke features for different clients, and are managing nearly 40 different apps now. 

We love Flutter and flavors have got us this far, but it would be great to see more support and DX for them!

**Comment by amorenew** [comment]
In my use case, while on web I don't want to load an image as a background but just a color to reduce page loading

**Comment by fulcri** [comment]
Up

**Comment by AmanMasipeddi** [comment]
+1

**Comment by maBarabas** [comment]
I would like to see assets that are bundled for only testing as well. Currently we use the assets feature to load databases written by previous versions of the app for testing migration code. We do not want these test databases embedded in the production build of the app.

Currently, the production build CI script will modify the pubspec to remove the test files before doing the build, but it would be better if there was a way to tell flutter that these assets are only for testing.

Related issue: https://github.com/flutter/flutter/issues/96514

**Comment by ueman** [comment]
Being able to use assets based on a flavor helps to reduce the app size. Currently we just bundle every asset for every flavor inside the app, which is very detrimental to the app size. With assets per flavor, we could get rid of this.

**Comment by giaur500** [comment]
Main problem is Flutter engine is not aware of existing Android flavors (or xcode schemes), `--flavor` build option only forwards flavor name to platform specific build script. 

Maybe you should add Flutter's own flavors support with optional association with native Android flavor/xcode scheme

**Comment by andrewkolos** [comment]
I think there is great opportunity to improve Flutter in regards to assets, and I would love to tackle this issue in particular.

Several folks have pointed out that designing this capability is not a straight-forward challenge. The method we come up with for associating assets with flavors needs to provide a good developer experience while also meshing well with other asset-related features we could add in the future, such as support for locale, platform, text-direction, dark/night mode, and so on.

Over the upcoming weeks, I plan on following the [Design Document](https://github.com/flutter/flutter/wiki/Design-Documents) process to document my research and make a proposal.

**Comment by giaur500** [comment]
So, is there any consensus how it should work? Which way is considered as preffered? As far as I understand, separate sections for each flavor in publspec is not an option.

**Comment by ueman** [comment]
That question will be answered in the upcoming design document andrewkolos mentioned.

**Comment by andrewkolos** [comment]
> So, is there any consensus how it should work? Which way is considered as preffered? As far as I understand, separate sections for each flavor in publspec is not an option.

Nope. Figuring out how it should work will be the primary goal of the design process. I will post a link to the design document here once a first-draft is ready, and anyone will be able to comment on it directly. In addition, it will be discussed live in a Dash Forum, which will be open to anyone with [contributor access](https://github.com/flutter/flutter/wiki/Contributor-access).

Regardless, if anyone has novel ideas for how they would like this feature to work, feel free to comment on this issue. There is no need to wait for a design document--your ideas would help inform it. There are many possibilities here.

**Comment by giaur500** [comment]
`Regardless, if anyone has novel ideas for how they would like this feature to work, feel free to comment on this issue. `

From my side, as end user: prefferable way is separate sections in publspec. It won't happen as it's been said above, so I'm waiting for any final idea that will be accepted by devs.

**Comment by andrewkolos** [comment]
I've written a proposal in the form of a design document, [Flavor-specific Assets](https://flutter.dev/go/flavor-specific-assets). This will be discussed in the next Dash Forum (August 8th, US time).

**Comment by giaur500** [comment]
Regarding to flavors, there is one more thing to be discussed. You probably know how it works with native Android apps. You can declare `main` resources and all flavors will use them, unless you overwrite any item for specific flavor. So, there's not need to duplicate assets, if you don't define any flavor specific one, default will be used.

As far as I understand, all ideas that have been considered here assume full set of assets for each flavor, so some assets must be actually duplicated.

**Comment by andrewkolos** [comment]
> As far as I understand, all ideas that have been considered here assume full set of assets for each flavor, so some assets must be actually duplicated.

Duplicate declaration (or any kind of duplicate work) is something the designs try to avoid. For all the pubspec-based proposals on the design doc (as of 8/2), the user would only need to declare every asset (or asset directory) once. Said another way:
* If an asset is common to all flavors, the user would simply declare it without a `flavor`.
* If an asset is unique to a flavor (or subset of flavors), the user would add the flavor(s) to that asset entry (or asset `group`, if choose to implement such a syntax).

I'll try to revise the document to make that aspect more clear. Also, anyone can comment on the doc using the comments feature in google docs, so feel free to add a comment to a part of the doc if it seems confusing.

**Comment by passsy** [comment]
Related https://github.com/flutter/flutter/issues/65065
